keps/0009-operator-toolkit.md

```diff
@@ -0,0 +1,296 @@
+
+### Tasks
+
+Tasks are a map of task name to a list of templates that should be executed in that step. With [Extensions and Bases](#extensions-and-bases), tasks can be represented in multiple forms, as long as the end result is a ready-to-run YAML manifest for a Kubernetes object.
+
+#### Files
+
+If a filename is specified, KUDO will execute a Go Template on the relevant filename, described more in detail in [Templates](#templates).
+
+#### Webhooks
+
+If a webhook is specified, KUDO will POST the listed webhook with a JSON body containing the full KUDO [execution state](#execution-state). The webhook **MUST** respond with the following:
+
+- Status code: 200
+- Content-type: application/json
+
+The body of this response **MUST** be a fully-qualified Kubernetes object in JSON form. If multiple objects are required, they **MUST** be wrapped in a Kubernetes List API object. The contents of a list object **CAN** be composed of multiple Kubernetes objects of different types.
+
+Webhooks **CAN** have side effects to control external state or to fetch external parameters for use in their templates, as long as the response is still a Kubernetes object.
+
+Webhook HTTP connections time out after 30 seconds, after which a Timeout event is added to the PlanExecution the task is running on.
+
+#### Resources vs. Patches
+
+KUDO additionally supports Kustomize for defining resources and patching for tasks. Kustomize is applied after any Go templating steps. This is useful for defining common bases for objects, or for extending frameworks as described in [Extensions and Bases](#extensions-and-bases).
+
+#### Task Application
+
+Once KUDO has assembled a full set of templates for a task, they will be applied using Kubernetes server-side apply at the same time.
+
+### Parameters
+
+Parameters are a map of Parameter structs, with the key being the name of the parameter. A default value can be set inside of a parameter, as well as a description. When no default is specified, the value is required. If an Instance is created without a required parameter, it will have an Error event added to it describing the validation failure. As described in [Execution State](#execution-state), they are wrapped into the `.Params` object for use by all templated objects. All keys, even arbitrary keys unused by KUDO, are wrapped into the `.Params` object for potential use by other templates.
+
+Parameters are intentionally open for extension for adding fields in the future for validation and more.
+
+### Templates
+
+A template is a standard Kubernetes manifest which **MAY** have additional Go Templating. These Go Templates include [Sprig](https://github.com/masterminds/sprig) to provide a standard library of functions familiar to Helm users. Additional values are available and are described in [Execution State](#execution-state).
+
+## Extensions and Bases
```

Comment:
I feel like extensions could also be phase 2, right? I mean it's good we thought it through, but we probably should implement the "simple" case fully, before we start to extend.

In general, even in programming I am not a big fan of inheritance because it's just too much context switching to be able to figure out how the overriden stuff pass into a bigger picture. I can totally see myself "cloning" existing package and "extending" it by changing some plans/adding new plans. But I guess then you lose the ability to update the version of underlying package. If packages are pretty much backward compatible and you can actually update, then I guess inheritance makes sense. I am bit undecided here :-) but I can easily be convinced.

Reply:
e.g. when someone changes name of the thing you were extending in next version, you lose backward compatibility ...

Reply:
Since the extension explicitly references a version of the `operator`, it'll have to be a choice to update to the new version.   This brings up a great check we should put into our `operator` test suite that forces a change in Version (how big a change TBD) when steps/plans/tasks/templates/parameters change.